proposals/0031-hlsl-vector-matrix-operations.md

```diff
@@ -156,12 +144,500 @@ vector<float, 3> ApplyNeuralMaterial(vector<half, 8> inputVector) {
+          typename MatrixBufferTy, DataType InputDT, DataType MatrixDT,
+          uint MatrixM, uint MatrixK, MatrixLayout MatrixLayout,
+          bool MatrixTranspose, typename BiasVectorBufferTy,
+          DataType BiasVectorDT>
+vector<OutputElTy, MatrixM>
+MulAdd(MatrixRefImpl<MatrixBufferTy, MatrixDT, MatrixM, MatrixK, MatrixLayout,
+                     MatrixTranspose>
+           Matrix,
+       Vector<InputElTy, InputElCount, InputDT> InputVector,
+       VectorRefImpl<BiasVectorBufferTy, BiasVectorDT> BiasVector) {
+
+  vector<OutputElTy, MatrixM> OutputVector;
+
+  details::__builtin_MatVecMulAdd(
+      /*out*/ OutputVector, details::IsUnsigned<OutputElTy>(), InputVector.Data,
+      details::IsUnsigned<InputElTy>(), InputDT, BUFFER_HANDLE(Matrix.Buffer),
+      Matrix.StartOffset, MatrixDT, MatrixM, MatrixK, MatrixLayout,
+      MatrixTranspose, Matrix.Stride, BUFFER_HANDLE(BiasVector.Buffer),
+      BiasVector.StartOffset, BiasVectorDT);
+
+  return OutputVector;
+}
+```
+
+## Function: OuterProductAccumulate
+
+Computes the outer product between column vectors and an **M**x**N** matrix is
+accumulated component-wise atomically (with device scope) in memory.
+
+The operation is equivalent to:
+
+> ResultMatrix = InputVector1 \* Transpose(InputVector2);
+
+Example:
+
+```c++
+RWByteAddressBuffer RWBuf;
+
+void Example(vector<half, 128> Input1, vector<half, 256> Input2) {
+  using namespace dx::linalg;
+
+  RWMatrixRef<DATA_TYPE_FLOAT16, 128, 256, MATRIX_LAYOUT_OUTER_PRODUCT_OPTIMAL>
+      matrix = {RWBuf, 0, 0};
+
+  OuterProductAccumulate(Input1, Input2, matrix);
+}
+```
+
+Conceptual API:
+
+```c++
+namespace dx {
+namespace linalg {
+
+void OuterProductAccumulate(vector<T, M> InputVector1,
+                            vector<T, N> InputVector2,
+                            RWMatrixRef<...> Matrix);
+
+} // namespace linalg
+} // namespace dx
+```
+
+Parameters:
+
+- `InputVector1` - the first vector, containing M elements. Element type must
+  be the same as InputVector2's.
+- `InputVector2` - the second vector, containing N elements. Element type must
+  be the same as InputVector1's.
+- `Matrix` - the destination matrix. The matrix dimensions must be MxN. The
+  `TRANSPOSE` parameter for the matrix must be `false`.
+
+Implementation:
+
+```c++
+namespace dx {
+namespace linalg {
+
+template <typename ElTy, int MatrixM, int MatrixN, DataType MatrixDT,
+          MatrixLayout MatrixLayout>
+void OuterProductAccumulate(
+    vector<ElTy, MatrixM> InputVector1, vector<ElTy, MatrixN> InputVector2,
+    RWMatrixRef<MatrixDT, MatrixM, MatrixN, MatrixLayout, false> Matrix) {
+  details::__builtin_OuterProductAccumulate(
+      InputVector1, InputVector2, BUFFER_HANDLE(Matrix.Buffer),
+      Matrix.StartOffset, MatrixDT, MatrixLayout, Matrix.Stride);
+}
+
+} // namespace linalg
+} // namespace dx
+```
+
+## Function: VectorAccumulate
+
+Accumulates the components of a vector component-wise atomically (with device
```

Comment:
Do you mean 'within device scope' ?

Reply:
I think both could be correct, but other references I find online tend to talk about an atomic operation having a particular scope, rather than taking place within a particular scope.